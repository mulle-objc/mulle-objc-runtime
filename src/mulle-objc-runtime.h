//
//  mulle_objc.h
//  mulle-objc-runtime
//
//  Created by Nat! on 10.07.16.
//  Copyright (c) 2016 Nat! - Mulle kybernetiK.
//  Copyright (c) 2016 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_objc_runtime_h__
#define mulle_objc_runtime_h__

/*
 * If you want to include <mulle-objc-runtime/mulle-objc-runtime.h> you
 * should be compiling with an Objective-C compiler, that defines the
 * following defines properly. Do not use .c  use .m with mulle-objc-runtime.h
 */

// catch this early
#if ! defined( __MULLE_OBJC_TPS__) && ! defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TPS__ or __MULLE_OBJC_NO_TPS__)"
#endif
#if defined( __MULLE_OBJC_TPS__) && defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TPS__ and __MULLE_OBJC_NO_TPS__)"
#endif
#if ! defined( __MULLE_OBJC_FCS__) && ! defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_FCS__ or __MULLE_OBJC_NO_FCS__)"
#endif
#if defined( __MULLE_OBJC_FCS__) && defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_FCS__ and __MULLE_OBJC_NO_FCS__)"
#endif
#if ! defined( __MULLE_OBJC_TAO__) && ! defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TAO__ or __MULLE_OBJC_NO_TAO__)"
#endif
#if defined( __MULLE_OBJC_TAO__) && defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TAO__ and __MULLE_OBJC_NO_TAO__)"
#endif

//
// these should have been defined by the compiler already, but for
// mulle-objc-runtime testing with gcc its convenient to add these if missing.
//
#ifndef __MULLE_OBJC_UNIVERSEID__
# define __MULLE_OBJC_UNIVERSEID__  0
#endif
#ifndef __MULLE_OBJC_UNIVERSENAME__
# define __MULLE_OBJC_UNIVERSENAME__  NULL
#endif

#if defined( __MULLE_OBJC_TPS__) && __MULLE_OBJC_UNIVERSEID__ != 0
# error "TPS can only be used in the default universe not in " #__MULLE_OBJC_UNIVERSENAME__
#endif


#include "include.h"

#include "mulle-metaabi.h"

#include "mulle-objc-atomicpointer.h"
#include "mulle-objc-builtin.h"
#include "mulle-objc-call.h"
#include "mulle-objc-class.h"
#include "mulle-objc-classpair.h"
#include "mulle-objc-class-convenience.h"
#include "mulle-objc-class-initialize.h"
#include "mulle-objc-class-lookup.h"
#include "mulle-objc-class-search.h"
#include "mulle-objc-class-struct.h"
#include "mulle-objc-csvdump.h"
#include "mulle-objc-fastclasstable.h"
#include "mulle-objc-fastmethodtable.h"
#include "mulle-objc-infraclass.h"
#include "mulle-objc-ivar.h"
#include "mulle-objc-ivarlist.h"
#include "mulle-objc-kvccache.h"
#include "mulle-objc-load.h"
#include "mulle-objc-metaclass.h"
#include "mulle-objc-method.h"
#include "mulle-objc-methodidconstants.h"
#include "mulle-objc-methodlist.h"
#include "mulle-objc-object.h"
#include "mulle-objc-objectheader.h"
#include "mulle-objc-object-convenience.h"
#include "mulle-objc-property.h"
#include "mulle-objc-propertylist.h"
#include "mulle-objc-retain-release.h"
#include "mulle-objc-signature.h"
#include "mulle-objc-super.h"
#include "mulle-objc-symbolizer.h"
#include "mulle-objc-taggedpointer.h"
#include "mulle-objc-try-catch-finally.h"
#include "mulle-objc-universe.h"
#include "mulle-objc-universe-class.h"
#include "mulle-objc-universe-exception.h"
#include "mulle-objc-universe-fail.h"
#include "mulle-objc-universe-global.h"
#include "mulle-objc-universe-struct.h"
#include "mulle-objc-uniqueid.h"
#include "mulle-objc-uniqueidarray.h"
#include "mulle-objc-version.h"
#include "mulle-objc-walktypes.h"



// add some functions to mulle-vararg for ObjC

#define mulle_vararg_count_ids( args, obj) \
   mulle_vararg_count_pointers( args, obj)
#define mulle_vararg_count_objects( args, obj) \
   mulle_vararg_count_pointers( args, obj)

#define mulle_vararg_next_id( args) \
   mulle_vararg_next_pointer( args, id)
#define mulle_vararg_next_object( args, type) \
   mulle_vararg_next_pointer( args, type)

#define mulle_vararg_push_id( args, value) \
   _mulle_vararg_push( args, id, value)
#define mulle_vararg_push_object( args, value) \
   mulle_vararg_push_id( args, value)

//
// add this before return of int main( argc, char **argc) when running
// named universes and leak checking
//
static inline void  mulle_objc_global_finish( void)
{
   struct _mulle_objc_universe   *universe;

   universe = mulle_objc_global_get_universe_inline( __MULLE_OBJC_UNIVERSEID__);
   if( universe && ! _mulle_objc_universe_is_default( universe) && universe->config.pedantic_exit)
      _mulle_objc_universe_release( universe);
}



#ifdef __has_include
# if __has_include( "_mulle-objc-runtime-versioncheck.h")
#  include "_mulle-objc-runtime-versioncheck.h"
# endif
#endif


#include <assert.h>
#include <stdarg.h>
#include <setjmp.h>

#endif

