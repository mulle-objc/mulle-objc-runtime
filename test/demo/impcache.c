#ifndef __MULLE_OBJC__
# define __MULLE_OBJC_NO_TPS__
# define __MULLE_OBJC_FCS__
# if defined( DEBUG) || ! defined( __OPTIMIZE__)
#  define __MULLE_OBJC_TAO__
# else
#  define __MULLE_OBJC_NO_TAO__
# endif
#endif


#include <mulle-objc-runtime/mulle-objc-runtime.h>
#include <stdio.h>



// uniqueid can be a methodid or superid!
// MEMO: this is a an old function, that it just used in tests its a little
//       strange, because it has no universe. Do not use otherwise.
//

MULLE_OBJC_RUNTIME_GLOBAL
struct _mulle_objc_cacheentry *
    demo_impcachepivot_fill( struct _mulle_objc_impcachepivot *cachepivot,
                             mulle_objc_implementation_t imp,
                             mulle_objc_uniqueid_t uniqueid,
                             unsigned int fillrate,
                             struct mulle_allocator *allocator);

struct _mulle_objc_cacheentry *
    demo_impcachepivot_fill( struct _mulle_objc_impcachepivot *cachepivot,
                             mulle_objc_implementation_t imp,
                             mulle_objc_uniqueid_t uniqueid,
                             unsigned int fillrate,
                             struct mulle_allocator *allocator)
{
   struct _mulle_objc_impcache     *icache;
   struct _mulle_objc_impcache     *old_cache;
   struct _mulle_objc_cache        *cache;
   struct _mulle_objc_cacheentry   *entry;

   assert( cachepivot);
   assert( imp);

   //
   // try to get most up to date value
   //
   for(;;)
   {
      cache = _mulle_objc_cachepivot_get_cache_atomic( &cachepivot->pivot);

      if( _mulle_objc_cache_should_grow( cache, fillrate))
      {
         old_cache = _mulle_objc_cache_get_impcache_from_cache( cache);
         icache    = _mulle_objc_impcache_grow_with_strategy( old_cache,
                                                              MULLE_OBJC_CACHESIZE_GROW,
                                                              allocator);

         // doesn't really matter, if this fails or succeeds we just try
         // again
         _mulle_objc_impcachepivot_swap( cachepivot,
                                         icache,
                                         old_cache,
                                         allocator);
         continue;
      }

      entry = _mulle_objc_cache_add_functionpointer_entry( cache,
                                                           (mulle_functionpointer_t) imp,
                                                           uniqueid);
      if( entry)
         break;
   }

   return( entry);
}



/* TODO:

   NSProxy              Object
   +-----------+       +------------+
   | forward:  |<----->|            |  [NSNotificationCenter self]
   +-----------+       +------------+

   NSProxy       
   +------------+       
   | forward:   |
   +------------+
   |            | 
   +------------+
   |            | [super ]
   +------------+

*/



#define N_CACHES  2

struct demo_class
{
   struct _mulle_objc_impcachepivot   cachepivot[ N_CACHES];
};


static void   demo_class_print_cache( struct demo_class *cls, unsigned int index)
{
   struct _mulle_objc_cache   *cache;
   mulle_objc_cache_uint_t    i, n;

   cache = _mulle_objc_cachepivot_get_cache_atomic( &cls->cachepivot[ index].pivot);
   n     = _mulle_objc_cache_get_size( cache);

   fprintf( stderr, "%d:\n", index);
   for( i = 0; i < n; i++)
   {
      fprintf( stderr, "   % 2d: [ 0x%08x, %p ]\n", 
                        i,
                        cache->entries[ i].key.uniqueid, 
                        (void *) cache->entries[ i].value.functionpointer);
   }
}


static void   *imp0( void *self, mulle_objc_methodid_t _cmd, void *_params)
{
   printf( "%s\n", __FUNCTION__);
   return( self);
}


static void   *imp1( void *self, mulle_objc_methodid_t _cmd, void *_params)
{
   printf( "%s\n", __FUNCTION__);
   return( self);
}


static void   *imp2( void *self, mulle_objc_methodid_t _cmd, void *_params)
{
   printf( "%s\n", __FUNCTION__);
   return( self);
}


static void   *imp3( void *self, mulle_objc_methodid_t _cmd, void *_params)
{
   printf( "%s\n", __FUNCTION__);
   return( self);
}


#define N_METHODS  4

static struct _mulle_objc_method   methods[ N_METHODS] = 
{
   {
      .descriptor =
      {
         0x10100,             // descriptor.methodid
         "@:",                // descriptor.signature
         "m0",                // descriptor.name
         0                    // descriptor.bits
      },
      .value = imp0
   },
   {
      .descriptor =
      {
         0x10110,             // descriptor.methodid
         "@:",                // descriptor.signature
         "m1",                // descriptor.name
         0                    // descriptor.bits
      },
      .value = imp1
   },
   {
      .descriptor =
      {
         0x10120,             // descriptor.methodid
         "@:",                // descriptor.signature
         "m2",                // descriptor.name
         0                    // descriptor.bits
      },
      .value = imp2
   },
   {
      .descriptor =
      {
         0x10130,             // descriptor.methodid
         "@:",                // descriptor.signature
         "m3",                // descriptor.name
         0                    // descriptor.bits
      },
      .value = imp3
   }
};


// MEMO: needed for linking will not be used in test
struct _mulle_objc_universe  *
   __register_mulle_objc_universe( mulle_objc_universeid_t universeid,
                                  char *universename)
{
   mulle_fprintf( stderr, "%s\n", __FUNCTION__);
   return( NULL);
//   struct _mulle_objc_universe    *universe;
//
//
//   universe = __mulle_objc_global_get_universe( universeid, universename);
//   if( ! _mulle_objc_universe_is_initialized( universe))
//   {
//      _mulle_objc_universe_bang( universe, 0, NULL, NULL);
//   }
//   return( universe);
}


static int   aba_free( void *aba,
                       void (*free)( void *, void *),
                       void *block,
                       void *owner)
{
   (*free)( block, owner);
}



int  main( int argc, char *argv[])
{
   struct demo_class             cls;
   unsigned int                  i;
   unsigned int                  j;
   unsigned int                  k;
   struct _mulle_objc_impcache   *icache;
   struct _mulle_objc_method     *method;
   struct mulle_allocator        allocator;
   mulle_objc_implementation_t   imp;

   mulle_fprintf( stderr, "%s\n", __FUNCTION__);

   // use this because the test allocator alone doesn't do abafree
   allocator         = mulle_stdlib_allocator;
   allocator.abafree = aba_free;

   memset( &cls, 0, sizeof( cls));

   for( i = 0; i < N_CACHES; i++)
   {
      icache = mulle_objc_impcache_new( 4, NULL, &allocator);

      // we don't have these callbacks
      icache->callback.call                       = 0;
      icache->callback.call_cache_collision       = 0;
      icache->callback.call_cache_miss            = 0;
      icache->callback.supercall                  = 0;
      icache->callback.supercall_cache_miss       = 0;
      icache->callback.supercall_cache_collision  = 0;
      // hmmm probably not needed either

      icache->callback.refresh_method_nofail      = 0; // _mulle_objc_class_refresh_method_nofail;
      icache->callback.refresh_supermethod_nofail = 0; // _mulle_objc_class_refresh_supermethod_nofail;


      // MEMO: as the cache will grow, first two methods will be evicted
      //       again
      if( _mulle_objc_impcachepivot_swap( &cls.cachepivot[ i],
                                          icache,
                                          NULL,
                                          &allocator))
         abort();

      for( j = 0; j < N_METHODS; j++)
      {
         method = &methods[ j];
          demo_impcachepivot_fill( &cls.cachepivot[ i],
                                   _mulle_objc_method_get_implementation( method),
                                   _mulle_objc_method_get_methodid( method),
                                   0,
                                   &allocator);
      }
   }

   for( i = 0; i < N_CACHES; i++)
   {
      for( j = 0; j < N_METHODS; j++)
      {
retry:
        // demo_class_print_cache( &cls, i);

         imp = _mulle_objc_impcachepivot_probe_inline( &cls.cachepivot[ i],
                                                       methods[ j].descriptor.methodid);
         if( imp)
            (*imp)( NULL, methods[ j].descriptor.methodid, NULL);
         else
         {
            for( k = 0; k < N_METHODS; k++)
               if( methods[ k].descriptor.methodid == methods[ j].descriptor.methodid)
               {
                  method = &methods[ k];
                  demo_impcachepivot_fill( &cls.cachepivot[ i],
                                           _mulle_objc_method_get_implementation( method),
                                           _mulle_objc_method_get_methodid( method),
                                           0,
                                           &allocator);
                  goto retry;     
               }
            // lookup method somehow and add to impcache
         }
      }
   }

   // get rid of it
   for( i = 0; i < N_CACHES; i++)
   {
      icache = _mulle_objc_impcachepivot_get_impcache_atomic( &cls.cachepivot[ i]);
      _mulle_objc_impcachepivot_swap( &cls.cachepivot[ i],
                                      NULL,
                                      icache,
                                      &allocator);
   }

   return( 0);
}
