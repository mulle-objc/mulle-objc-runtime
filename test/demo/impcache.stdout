imp0
imp1
imp2
imp3
imp0
imp1
imp2
imp3
